Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,427 @@
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <memory>
+#include <utility>
+
+// #define _ENABLE_ACTS_ANY_DEBUG
+
+#if defined(_ENABLE_ACTS_ANY_DEBUG)
+#include <iomanip>
+#include <iostream>
+#define _ACTS_ANY_DEBUG(x) std::cout << x << std::endl;
+#define _ACTS_ANY_DEBUG_BUFFER(s, b)   \
+  do {                                 \
+    std::cout << "" << s << ": 0x";    \
+    for (char c : b) {                 \
+      std::cout << std::hex << (int)c; \
+    }                                  \
+    std::cout << std::endl;            \
+  } while (0)
+#else
+#define _ACTS_ANY_DEBUG(x)
+#define _ACTS_ANY_DEBUG_BUFFER(s, b)
+#endif
+
+namespace Acts {
+
+/// Small opaque cache type which uses small buffer optimization
+template <size_t SIZE>
+class AnyBase {
+ public:
+  template <typename T, typename... Args>
+  static AnyBase make(Args&&... args) {
+    AnyBase cache{};
+
+    using U = std::decay_t<T>;
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    cache.m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (cache.m_data.data()) U(std::forward<Args>(args)...);
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      *reinterpret_cast<U**>(cache.m_data.data()) = heap;
+    }
+
+    return cache;
+  }
+
+  template <typename T, typename... Args>
+  explicit AnyBase(std::in_place_type_t<T> /*unused*/, Args&&... args) {
+    using U = std::decay_t<T>;
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::forward<Args>(args)...);
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      *reinterpret_cast<U**>(m_data.data()) = heap;
+    }
+  }
+
+#if defined(_ENABLE_ACTS_ANY_DEBUG)
```

Comment:
Is this really necessary, the `_ACTS_ANY_DEBUG` macro expands to nothing if `_ENABLE_ACTS_ANY_DEBUG` is not defined, right?

Reply:
There's some extra looping etc that I want to avoid if possible.